modules/distribution-zones/tech-notes/rebalance.md

```diff
@@ -55,18 +55,55 @@ But for the new one we have an idea, which doesn't need the metastore at all:
-- On every PrimaryReplica reelection by PD it must send the new RebalanceRequest to PrimaryReplica, if pending key is not empty. 
-- On every leader reelection (for the leader oriented protocols) inside the replication group - leader send leaderElected event to PrimaryReplica, which force PrimaryReplica to send RebalanceRequest to the replication group leader again.
+- On every new PD leader elected - it must check the direct value (not the locally cached one) of `zoneId.assignment.pending`/`zondeId.assignment.cancel` (the last one always wins, if exists) keys and send `RebalanceRequest`/`CancelRebalanceRequest` to needed PrimaryReplicas and then listen updates from the last revision of this key.
+- On every PrimaryReplica reelection by PD it must send the new `RebalanceRequest`/`CancelRebalanceRequest` to PrimaryReplica, if pending/cancel (cancel always wins, if filled) key is not empty. 
+- On every leader reelection (for the leader oriented protocols) inside the replication group - leader send leaderElected event to PrimaryReplica, which force PrimaryReplica to send RebalanceRequest/CancelRebalanceRequest to the replication group leader again.
+
+## Cancel an ongoing rebalance process if needed
+Sometimes we must cancel the ongoing rebalance:
+- We can receive an unrecoverable error from replication group during the current rebalance
+- We can decide to cancel it manually
 
+![](images/cancelRebalance.svg)
+
+### 1. Put rebalance intent to *.cancel key
+For the purpose of persisting for cancel intent, we must save the (oldTopology, newTopology) pair of peers lists to `zoneId.assignment.cancel` key.
+Also, every invoke with update of `*.cancel` key must be enriched by revision of the pending key, which must be cancelled:
+```
+    if(zoneId.assignment.pending.revision == inputRevision):
+        zoneId.assignment.cancel = cancelValue
+        return true
+    else:
+        return false
+```
+It's needed to prevent the race, between the rebalance done and cancel persisting, otherwise we can try to cancel the wrong rebalance process.
 
+### 2. PrimaryReplica->ReplicationGroup cancel protocol
+When PrimaryReplica send `CancelRebalanceRequest(oldTopology, newTopology)` to the ReplicationGroup following cases available:
+- Replication group has ongoing rebalance oldToplogy->newTopology. So, it must be cancelled and cleanup for the configuration state of replication group to  oldTopology must be executed.
+- Replication group has no ongoing rebalance and currentTopology==oldTopology. So, nothing to cancel, return success response.
+- Replication group has no ongoing rebalance and currentTopology==newTopology. So, cancel request can't be executed, return the response about it. Result recipient of this response (placement driver) must log this fact and do the same routine for usual rebalanceDone.
```

Comment:
What if after sending `CancelRebalanceRequest` the placement driver finds out that some of replication groups have finished rebalance (currentTopology==newTopology) and some have not?

Reply:
How does it possible, if any rebalance touches only one distribution zone, so, only one replication group?

Reply:
Seems I missed this moment, my fault. Question withdrawn.